File: hdv-foundation/nodes/persona.ts
/**
 * nodes/persona.ts — the ephemeral persona lifecycle: spawn -> execute -> terminate.
 *
 * A persona is the smallest unit of work in the matrix. It is created, does exactly one
 * job, and is destroyed. Personas never persist. Each is conceptually tied to a model.
 *
 * When HDV_PERSONA_INFERENCE=1 (or live:true), executePersona drives a REAL Ollama
 * completion and folds the text into the score. Otherwise it keeps the deterministic
 * offline pseudo-score (tests / no model).
 */
import { execFileSync } from 'node:child_process';
import { randomUUID } from 'node:crypto';
import type { AgentRole } from '../config/routing_schema.js';
import { MODEL_SIZE } from './constants.js';

export type PersonaState = 'SPAWNED' | 'EXECUTING' | 'TERMINATED';

export interface Persona {
  id: string;
  owner: AgentRole;
  nodeId: string;
  modelSize: string;
  state: PersonaState;
  spawnedAt: number;
  terminatedAt?: number;
}

export interface PersonaExecution {
  personaId: string;
  input: Record<string, unknown>;
  score: number;
  durationMs: number;
  /** Provenance when a live model ran the persona. */
  backend?: 'stub' | 'ollama';
  modelId?: string;
  text?: string;
}

export interface ExecutePersonaOptions {
  /** Force live Ollama inference for this call. */
  live?: boolean;
}

/** SPAWN — create an ephemeral persona bound to a node of a given Big AI. */
export function spawnPersona(owner: AgentRole, nodeId: string): Persona {
  return {
    id: `persona_${randomUUID()}`,
    owner,
    nodeId,
    modelSize: MODEL_SIZE,
    state: 'SPAWNED',
    spawnedAt: Date.now(),
  };
}

/**
 * EXECUTE — run the persona's single job.
 * Live path (HDV_PERSONA_INFERENCE=1): Ollama generate → score from generated text.
 * Offline path: deterministic pseudo-score (reproducible tests, no model required).
 */
export function executePersona(
  persona: Persona,
  input: Record<string, unknown>,
  options: ExecutePersonaOptions = {},
): PersonaExecution {
  if (persona.state === 'TERMINATED') {
    throw new Error(`persona ${persona.id} already terminated — cannot execute`);
  }
  persona.state = 'EXECUTING';
  const live =
    options.live === true ||
    (options.live !== false &&
      ['1', 'true', 'yes', 'on'].includes((process.env.HDV_PERSONA_INFERENCE ?? '').trim().toLowerCase()));

  if (live) {
    try {
      return executePersonaOllama(persona, input);
    } catch (err) {
      // Fall through to stub so a transient Ollama blip never bricks routing.
      if (process.env.HDV_PERSONA_INFERENCE_STRICT === '1') throw err;
    }
  }

  const score = pseudoScore(JSON.stringify(input) + persona.id);
  return { personaId: persona.id, input, score, durationMs: 1, backend: 'stub', modelId: 'stub-1' };
}

/** TERMINATE — destroy the persona. Ephemeral by contract; no reuse afterward. */
export function terminatePersona(persona: Persona): Persona {
  persona.state = 'TERMINATED';
  persona.terminatedAt = Date.now();
  return persona;
}

/** Live Ollama persona: one short completion, score folded from the generated text. */
function executePersonaOllama(persona: Persona, input: Record<string, unknown>): PersonaExecution {
  const started = Date.now();
  const baseRaw =
    process.env.PERSONAMATRIX_OLLAMA_URL ||
    process.env.OLLAMA_HOST ||
    process.env.HDV_LLM_BASE_URL ||
    'http://127.0.0.1:11434';
  const base = baseRaw.replace(/\/v1\/?$/, '').replace(/\/$/, '');
  const root = base.includes('://') ? base : `http://${base}`;
  const model =
    process.env.PERSONAMATRIX_MODEL_ID || process.env.HDV_LLM_MODEL || 'llama3.2:3b';
  const intent =
    typeof input.intent === 'string' && input.intent.trim()
      ? input.intent.trim()
      : JSON.stringify(input);
  const prompt =
    `You are ephemeral persona ${persona.id} under ${persona.owner}. ` +
    `Reply in ONE short sentence (max 40 words) for this task:\n${intent}`;
  const payload = JSON.stringify({
    model,
    prompt,
    stream: false,
    options: { num_predict: 48, temperature: 0.4 },
  });
  const out = execFileSync(
    'curl',
    ['-sS', '--max-time', '120', '-X', 'POST', `${root}/api/generate`, '-H', 'Content-Type: application/json', '-d', payload],
    { encoding: 'utf8', maxBuffer: 2 * 1024 * 1024 },
  );
  const parsed = JSON.parse(out) as { response?: string };
  const text = (parsed.response ?? '').trim() || `(empty ollama response for ${persona.id})`;
  const score = pseudoScore(text);
  return {
    personaId: persona.id,
    input,
    score,
    durationMs: Date.now() - started,
    backend: 'ollama',
    modelId: model,
    text,
  };
}

/** Deterministic 0..1 pseudo-score (FNV-1a hash normalized). */
function pseudoScore(seed: string): number {
  let h = 0x811c9dc5;
  for (let i = 0; i < seed.length; i++) {
    h ^= seed.charCodeAt(i);
    h = Math.imul(h, 0x01000193);
  }
  return ((h >>> 0) % 10000) / 10000;
}

File: hdv-foundation/personamatrix/persona.py
"""persona.py -- the ephemeral persona loop driven by the filter_director.

Lifecycle, mirrored from the TypeScript nodes layer: spawn -> execute -> terminate.
Each persona is conceptually tied to a 7B model. The `filter_director` applies the
tuning params from config/filters.json (intensity, waveSpeed, shift, ...) to shape a
persona's ephemeral behavior, then tears it down.

Standard library only (Phase 1).
"""
from __future__ import annotations

import time
import uuid
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Dict, List, Optional

from .config_loader import filter_params, load_matrix
from .model_backend import (
    GenerationRequest,
    ModelBackend,
    default_backend,
    deterministic_seed,
)


class PersonaState(str, Enum):
    SPAWNED = "SPAWNED"
    EXECUTING = "EXECUTING"
    TERMINATED = "TERMINATED"


@dataclass
class Persona:
    owner: str
    node_id: str
    model_size: str = "7B"
    id: str = field(default_factory=lambda: f"persona_{uuid.uuid4()}")
    state: PersonaState = PersonaState.SPAWNED
    spawned_at: float = field(default_factory=time.time)
    terminated_at: Optional[float] = None


@dataclass
class PersonaExecution:
    persona_id: str
    score: float
    output: Dict[str, Any]


def spawn(owner: str, node_id: str) -> Persona:
    """SPAWN -- create an ephemeral persona bound to a node of a Big AI."""
    matrix = load_matrix()
    model_size = matrix.get("topology", {}).get("modelSize", "7B")
    return Persona(owner=owner, node_id=node_id, model_size=model_size)


def execute(
    persona: Persona,
    payload: Dict[str, Any],
    filters: Optional[Dict[str, float]] = None,
    backend: Optional[ModelBackend] = None,
) -> PersonaExecution:
    """EXECUTE -- run the persona's single job through the model backend + filter transform.

    ``backend`` is the pluggable inference seam (see ``model_backend``). When omitted, the
    process-wide deterministic ``StubBackend`` is used, so behavior is byte-for-byte
    backward compatible with the historical persona loop. Pass a ``TransformersBackend`` to
    drive a real 7B-class model on GPU.
    """
    if persona.state == PersonaState.TERMINATED:
        raise RuntimeError(f"persona {persona.id} already terminated -- cannot execute")
    persona.state = PersonaState.EXECUTING
    f = filters if filters is not None else filter_params()
    active_backend = backend if backend is not None else default_backend()

    result = active_backend.generate(
        GenerationRequest(persona_id=persona.id, payload=payload, filters=f)
    )

    output: Dict[str, Any] = {
        "owner": persona.owner,
        "node_id": persona.node_id,
        "applied_filters": f,
        "backend": result.backend,
        "model_id": result.model_id,
    }
    if result.text:
        output["text"] = result.text

    return PersonaExecution(persona_id=persona.id, score=result.score, output=output)


def terminate(persona: Persona) -> Persona:
    """TERMINATE -- destroy the persona. Ephemeral by contract; no reuse."""
    persona.state = PersonaState.TERMINATED
    persona.terminated_at = time.time()
    return persona


def filter_director(
    owner: str,
    node_id: str,
    payloads: List[Dict[str, Any]],
    filters: Optional[Dict[str, float]] = None,
    backend: Optional[ModelBackend] = None,
) -> List[PersonaExecution]:
    """Direct a full persona loop for a batch of payloads.

    For each payload: spawn -> execute (through the backend + filters) -> terminate. Returns
    one execution record per payload. This is the canonical persona loop for the matrix.
    ``backend`` defaults to the deterministic StubBackend (see ``execute``); pass a real
    backend once and it is shared across the whole batch (loaded once, reused per persona).
    """
    results: List[PersonaExecution] = []
    for payload in payloads:
        p = spawn(owner, node_id)
        results.append(execute(p, payload, filters, backend=backend))
        terminate(p)
    return results


def _seed(text: str) -> float:
    """Deterministic 0..~10 float seed from text (FNV-1a normalized).

    Retained for backward compatibility; delegates to ``model_backend.deterministic_seed``.
    """
    return deterministic_seed(text)
